airflow/settings.py

```diff
@@ -51,6 +52,34 @@
 log.info("Configured default timezone %s" % TIMEZONE)
 
 
+class RetryingQuery(Query):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        try:
+            self.max_tries = conf.getint('core', 'SQL_ALCHEMY_STATEMENT_MAX_RETRIES')
+        except conf.AirflowConfigException:
+            self.max_tries = 10
+
+        try:
+            self.max_retry_time_seconds = conf.getint('core', 'SQL_ALCHEMY_STATEMENT_MAX_RETRY_SECONDS')
+        except conf.AirflowConfigException:
+            self.max_retry_time_seconds = 30
+
+    def __iter__(self):
+        try_number = 1
+        while try_number <= self.max_tries:
+            try:
+                return super().__iter__()
+
+            except Exception as ex:
+                self.log.warn(f'Try {try_number}/{self.max_tries} failed to perform db action.', ex)
```

Comment:
log is defined in the global scope, don't think it's in self here.

Reply:
should we consider passing exc_info() here? motivation being, it was a bit hard to read the exception tied to the retry when we were parsing the logs last night

https://docs.python.org/3/library/logging.html#logging.Logger.debug

Reply:
sgtm!

Reply:
What is `exc_info()`?

The whole entire exception is forwarded to the logger.

Reply:
Here is a snippet:

```
[2019-03-29 03:54:54,885] {{logging_mixin.py:106}} WARNING - --- Logging error ---
[2019-03-29 03:54:54,885] {{logging_mixin.py:106}} WARNING - Traceback (most recent call last):
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/contrib/operators/kubernetes_pod_operator.py", line 133, in execute
    get_logs=self.get_logs)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/contrib/kubernetes/pod_launcher.py", line 88, in run_pod
    time.sleep(1)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/models/__init__.py", line 1371, in signal_handler
    raise AirflowException("Task received SIGTERM signal")
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - airflow.exceptions.AirflowException: Task received SIGTERM signal
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - 
During handling of the above exception, another exception occurred:
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - Traceback (most recent call last):
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/utils/db.py", line 88, in wrapper
    return func(*args, **kwargs)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/models/__init__.py", line 1394, in _run_raw_task
    result = task_copy.execute(context=context)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/contrib/operators/kubernetes_pod_operator.py", line 145, in execute
    raise AirflowException('Pod Launching failed: {error}'.format(error=ex))
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - airflow.exceptions.AirflowException: Pod Launching failed: Task received SIGTERM signal
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - 
During handling of the above exception, another exception occurred:
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING - Traceback (most recent call last):
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/logging/__init__.py", line 1034, in emit
    msg = self.format(record)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/logging/__init__.py", line 880, in format
    return fmt.format(record)
[2019-03-29 03:54:54,886] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/logging/__init__.py", line 619, in format
    record.message = record.getMessage()
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/logging/__init__.py", line 380, in getMessage
    msg = msg % self.args
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING - TypeError: not all arguments converted during string formatting
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING - Call stack:
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/bin/airflow", line 32, in <module>
    args.func(args)
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/bin/cli.py", line 528, in run
    _run(args, dag, ti)
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/bin/cli.py", line 443, in _run
    pool=args.pool,
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING -   File "/usr/local/lib/python3.7/site-packages/airflow/utils/db.py", line 90, in wrapper
    log.warn(f'Try {try_number}/{max_tries} failed to perform db action.', e)
[2019-03-29 03:54:54,887] {{logging_mixin.py:106}} WARNING - Message: 'Try 1/10 failed to perform db action.'
Arguments: (AirflowException('Pod Launching failed: Task received SIGTERM signal'),)
```

The issue here is that the exception gets formatted by the logger.

Reply:
ah shoot this LoggingMixin is slightly different than the pure logger module:

we need to pass a truthy value for `exc_info` as the exception instance isn't handled correctly by default.

Reply:
oh didn't mean to have you look at debug, it's just the only place where exc_info is described in the documentation